### PR TITLE
Initial introduction of delayed config concepts (Property, Provider, etc) for `QuarkusPluginExtension`, `QuarkusBuild` and `QuarkusDev`

### PR DESCRIPTION
**Pull request by sebersole**
Work from discussions at https://github.com/quarkusio/quarkus/discussions/25511

### CODE REVIEW

**Comment by sebersole** [comment]
This first commit simply migrates things directly to use deferred configuration concepts (`DirectoryProperty` instead of `File`, e.g.).  

I think some of the properties unnecessary probably.  And I would like to understand better about the single-source-directory limitation some more.  But those are other discussions.  I just wanted to focus on the configuration changes.

**Comment by glefloch** [comment]
Thanks @sebersole this is great! What is the minimal supported gradle version for providers? 
Regarding the single source I think this not a limitation anymore, the model we build supports a list of source directories IIRC.

**Comment by aloubyansky** [comment]
@sebersole is there anything you were planning to add to this PR before making it ready for review/merging? We don't run the full CI for drafts.

**Comment by aloubyansky** [comment]
Yes, `@Option` looks like too much for these. I do recall an issue with the `workingDir` some years ago. There is a comment about possible issues in Kotlin projects. We should make sure they still work after this refactoring.

**Comment by sebersole** [comment]
> Yes, `@Option` looks like too much for these. 

This ties-in with https://github.com/quarkusio/quarkus/discussions/25511#discussioncomment-2768454.  

If these really are not injectable, and are not really settable as I suspect, then objections to dropping these as properties?


> I do recall an issue with the `workingDir` some years ago. There is a comment about possible issues in Kotlin projects. We should make sure they still work after this refactoring.

I'm fairly confident that this is related to lack of deferrable configuration

**Comment by aloubyansky** [comment]
All your suggestions here seem reasonable to me @sebersole WDYT @glefloch?
I'd probably prefer separate PRs though unless they have to be combined for some reason.

**Comment by aloubyansky** [comment]
> > Yes, `@Option` looks like too much for these.
> 
> This ties-in with [#25511 (comment)](https://github.com/quarkusio/quarkus/discussions/25511#discussioncomment-2768454).
> 
> If these really are not injectable, and are not really settable as I suspect, then objections to dropping these as properties?

Let's remove those `@Option`s.

> > I do recall an issue with the `workingDir` some years ago. There is a comment about possible issues in Kotlin projects. We should make sure they still work after this refactoring.
> 
> I'm fairly confident that this is related to lack of deferrable configuration

Ah, indeed, looks like it could be.

**Comment by aloubyansky** [comment]
@sebersole may I ask you to please collect or at least highlight to us some changes that could cause backwards incompatibility issues even if you think they are unlikely to affect our current users based on our imagination how the plugin is used? I mean changes like removal of `@Option`? It won't hurt mentioning them in our release notes, just in case.

**Comment by aloubyansky** [comment]
`mvn -Dquickly` in the `gradle` module will take care of the failure.

**Comment by sebersole** [comment]
> @sebersole may I ask you to please collect or at least highlight to us some changes that could cause backwards incompatibility issues even if you think they are unlikely to affect our current users based on our imagination how the plugin is used? I mean changes like removal of `@Option`? It won't hurt mentioning them in our release notes, just in case.

Absolutely

**Comment by sebersole** [comment]
Sorry, I was in the wrong "gradle" directory

**Comment by sebersole** [comment]
> @sebersole may I ask you to please collect or at least highlight to us some changes that could cause backwards incompatibility issues even if you think they are unlikely to affect our current users based on our imagination how the plugin is used? I mean changes like removal of `@Option`? It won't hurt mentioning them in our release notes, just in case.

If you mean at the moment, I actually think nothing changes.  I changed the types of properties, but left the old accessors in place which simply delegate to the new ones.  If that is not accurate, I can add back whichever y'all want.

If you mean moving forward, based on https://github.com/quarkusio/quarkus/pull/25615#issuecomment-1129166712 and the other discussions... That depends on what type of changes y'all want to accept.

I think you should consider dropping all of the non-property properties, but that's ultimately a decision you have to make.  It is important there to consider the "impact" - changing those values has no effect or "wont work" based on what I saw, so in that respect I don't think there are compatibility concerns involved in dropping those.  However, a build script might just use them which would cause an error if you dropped them.  Assuming they were are not meant to be settable by users however, I'm not sure the error is a bad thing.

That would have "an impact" in that if

**Comment by sebersole** [comment]
> Are those things you really enforce? They always seem redundant to me, especially for Java Bean properties.

What's even more strange is that methods I did not even touch fail for some of the same reasons.  E.g.

```
    /**
     * Returns the last file from the specified {@link FileCollection}.
     * Needed for the Scala plugin.
     */
    public static File getLastFile(FileCollection fileCollection) {
```

leads to

```
/home/sebersole/projects/quarkus/quarkus/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/extension/QuarkusPluginExtension.java:251: warning: no @param for fileCollection
    public static File getLastFile(FileCollection fileCollection) {
                       ^
/home/sebersole/projects/quarkus/quarkus/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/extension/QuarkusPluginExtension.java:251: warning: no @return
    public static File getLastFile(FileCollection fileCollection) {
                       ^
```

**Comment by glefloch** [comment]
The provider API seems to be present for quite a long time. The current minimal version we support is 6.1 so it should be good.

Compilation warnings are generated by gradle (the maven command runs gradle), I think the `com.gradle.plugin-publish` plugin requires the public API to be documented. 

For next steps, I think we should first display a warning message and deprecate the setter to see if user are impacted by this. WDYT ?

**Comment by glefloch** [comment]
BTW, those warning won't fail the build but we want to fix them in a further step.

File: devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusBuild.java [hunk 1]
```diff
@@ -54,16 +73,21 @@ public QuarkusBuild nativeArgs(Action<Map<String, ?>> action) {
         return this;
     }
 
-    @Optional
     @Input
-    public List<String> getIgnoredEntries() {
+    @Optional
```
Comment (glefloch) [inline devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusBuild.java:77]: This optional can be removed too right ?

**Review by glefloch — APPROVED** [review]

**Comment by sebersole** [comment]
> For next steps, I think we should first display a warning message and deprecate the setter to see if user are impacted by this. WDYT ?

That's reasonable.

**Comment by sebersole** [comment]
> BTW, those warning won't fail the build but we want to fix them in a further step.

The javadoc errors?  They did fail the build.  I could not run the reformatting because of it.

**Comment by sebersole** [comment]
> Let's remove those `@Option`s.

@aloubyansky I just sent a PR[1] for this part of the discussion addressing those directory-related properties.

For the moment I left the `@Option` there.  It gets logged as deprecated just the same.  And we can drop it whenever y'all decide to drop them.  Or let me know there and I can remove them.


[1] https://github.com/quarkusio/quarkus/pull/25636

**Comment by sebersole** [comment]
How can I run these integration tests locally?

**Comment by sebersole** [comment]
What do you think about moving these under `devtools/gradle`?  As a noob, I had no idea these other tests existed.  Keeping them with the code they test would have helped, and is pretty conventional for Gradle projects.

I must be missing something though as the command is Maven.

**Comment by aloubyansky** [comment]
We had problems setting those tests up in the plugin module. Then we had them as Gradle project under the integration-tests. And also had issues (I think one of those was the time it took for the main Maven build process that drives the whole thing). @glefloch and @gsmet could probably give you more detailed info on that.

**Comment by aloubyansky** [comment]
Basically, the plugin hasn't been published yet at the time the tests where run. IIRC, there was an option "plugin under test" or something we had to configure but we still had some issues with that.

**Comment by sebersole** [comment]
"plugin under test" - yes, that is a TestKit setting.  I believe for that to work the plugin and tests cannot be split like you have.  Or did you mean it was a problem even before splitting the tests into separate project?

**Comment by aloubyansky** [comment]
Right, we had it and we still decided to separate them. I don't recall all the reasons right now. Perhaps @glefloch does.

**Comment by sebersole** [comment]
Really all I want is to be able to run these tests in the IDE.  I have no idea how to do that atm.  Pointers?

**Comment by aloubyansky** [comment]
> Really all I want is to be able to run these tests in the IDE. I have no idea how to do that atm. Pointers?

I just did it from Eclipse. Although at least some of the tests launch a gradle build in a separate process. E.g. to test `quarkusDev` and then check the responses from the endpoints. So the Gradle wrapper included in the module is used for that.

Generally speaking though, the Quarkus devs have been struggling to run tests from an IDE w/o doing some tricks like closing some modules due to the massive codebase. And still my impression is most are running the tests from the command line.

**Comment by sebersole** [comment]
> `cd integration-tests/gradle` and then either `mvn clean test` or e.g. `mvn clean test -Dtest=ResourcesInBuildStepsDevModeTest`

I cannot get this to work.  Locally, I even rolled back to a clean `main` branch and they still fail for me.  The errors seem like it is still seeing older versions though.  After changing the plugin code, what do I need to do to be able to run its tests?

**Comment by aloubyansky** [comment]
`mvn -Dquickly` the plugin first and then run the integration tests.

**Comment by sebersole** [comment]
Thanks @aloubyansky.  So I can understand, what does `mvn -Dquickly` in the plugin do?

Ok, if I update the README with some of these "noob gotchas"?

**Comment by glefloch** [comment]
It assemble the plugin, format source code and publish it locally.

**Comment by aloubyansky** [comment]
Basically, it's a quick `clean install` with formatting. See https://github.com/quarkusio/quarkus/blob/main/pom.xml#L165-L180

**Comment by sebersole** [comment]
`install` being equivalent  to `publishToMavenLocal`, right?

People who might want to help with this plugin may not necessarily know Maven.  E.g., its been YEARS since I have had to use Maven in any form of anger.

I'd also like to, with y'alls approval ofc, create a README in the 2 plugin projects as well.  They would be a better landing spot from their pages on the plugin portal.

**Comment by glefloch** [comment]
> install being equivalent to publishToMavenLocal, right?

Yes this is equivalent

Definitely ok for a README in each project. I can help if you want.

**Comment by sebersole** [comment]
I'm still missing something- 

1. I am on the latest of the `main` branch
2. I `cd` into `devtools/gradle/gradle-application-plugin/` and run `../../mvnw -Dquickly`
3. `cd` into `integration-tests/gradle` and run `../../mvnw clean test`

I get test failures

P.S. I use the bundled `mvnw` assuming it works like `gradlew`.  I do not have Maven installed, nor want to.  
P.S.S. I also "building the plugin" from `devtools/gradle/` - no diff

**Comment by sebersole** [comment]
> Definitely ok for a README in each project. I can help if you want.

Its low hanging fruit and something I can do after I actually get it working and understand :rofl: 

I don't want to distract you from your "real work".  Of course, I'm not going to tackle you if you really want to write them lol

**Comment by glefloch** [comment]
What kind of failure do you have? you may need to first run a `./mvnw -Dquickly` in the root directory to get all extensions published

**Comment by sebersole** [comment]
I cannot even run that.  Running `./mvnw -Dquickly` from the project root directory fails -

```
...
[INFO] Quarkus - AWT - Deployment ......................... SUCCESS [  0.305 s]
[INFO] Quarkus - Dev tools - Project Core Extension Codestarts SUCCESS [  5.967 s]
[INFO] Quarkus - BOM - Descriptor JSON .................... FAILURE [  1.129 s]
[INFO] Quarkus - Maven Plugin ............................. SKIPPED
...
```

**Comment by jskillin-idt** [comment]
For what it's worth, I've been doing `./mvnw -Dquickly && ./mvnw -f integration-tests/gradle/ install` from the root of the repo and it's been working for me, but I move the .m2 directory out of the way, and I also move my ~/.gradle/init.gradle and ~/.gradle/gradle.properties to ensure none of my flags mess with the build. In particular, "org.gradle.caching=true" causes integration test failures because the strings "SUCCESS" and "FROM-CACHE" are not equal ;)

**Comment by sebersole** [comment]
> For what it's worth, I've been doing `./mvnw -Dquickly && ./mvnw -f integration-tests/gradle/ install` from the root of the repo and it's been working for me

Thanks @jskillin-idt I'll give that a try.

> but I move the .m2 directory out of the way

You mean delete it completely?  Personally I have found relying on SNAPSHOTs in mavenLocal is usually not a good idea; at least from Gradle, not sure about Maven.  Unless the project configures not caching snapshots.

> I also move my ~/.gradle/init.gradle and ~/.gradle/gradle.properties to ensure none of my flags mess with the build. In particular, "org.gradle.caching=true" causes integration test failures because the strings "SUCCESS" and "FROM-CACHE" are not equal ;)

Not a problem for me.  Best practice is to set that on the project anyway, depending on whether that project wants build caching.  E.g. https://github.com/hibernate/hibernate-orm/blob/main/settings.gradle#L210

As for build-caching and tests, I have no idea about what does that check.  But considering `FROM-CACHE` is essentially the same as `SUCCESS`, can't that code just be changed?

This brings up a question for me... is a goal here to make the Quarkus tasks cacheable?  Definitely a separate discussion, but curious about the road map

**Comment by sebersole** [comment]
Trying `./mvnw clean install -Dquickly` to see if cleaning helps.

After that I guess I'll try deleting `~/.m2`; and as last resort re-cloning the repo

**Comment by jskillin-idt** [comment]
> You mean delete it completely? Personally I have found relying on SNAPSHOTs in mavenLocal is usually not a good idea; at least from Gradle, not sure about Maven. Unless the project configures not caching snapshots.

No, just move it out of the way. I have a few reasons for doing this, and most of them come down to being pessimistic about builds I haven't fully investigated, and also it just guarantees that there's no sneaky bits (flags, artifacts, whatever) creating untested combinations for the build tool.

> Not a problem for me. Best practice is to set that on the project anyway, depending on whether that project wants build caching. E.g. https://github.com/hibernate/hibernate-orm/blob/main/settings.gradle#L210

I live on the edge. :) Gradle's performance optimizations were hard to look past once I tasted how fast they run, and I have actually had no issues with them given that they're written very defensively.

> As for build-caching and tests, I have no idea about what does that check. But considering `FROM-CACHE` is essentially the same as `SUCCESS`, can't that code just be changed?

I guess it could! It was such a non-issue for me that it didn't even occur to me that I should write an issue for it.

> This brings up a question for me... is a goal here to make the Quarkus tasks cacheable? Definitely a separate discussion, but curious about the road map

I'm not in charge of the roadmap, but I definitely think this would be a big quality of life improvement. Like I said earlier, the performance improvements from Gradle's optimization features are addicting. My understanding is Gradle will cache stuff naturally if you use its API for task inputs/outputs, so I think first focusing on refactoring the plugin to use more of Gradle's API, like you're doing here, will make it easier to then make that jump.

**Comment by sebersole** [comment]
> I'm not in charge of the roadmap, but I definitely think this would be a big quality of life improvement. Like I said earlier, the performance improvements from Gradle's optimization features are addicting. 

For sure.  That's why I ask.  I saw pretty significant performance improvements in the Hibernate build when I applied all of these techniques.

> My understanding is Gradle will cache stuff naturally if you use its API for task inputs/outputs, so I think first focusing on refactoring the plugin to use more of Gradle's API, like you're doing here, will make it easier to then make that jump.

Not really.  At least it won't make use of the "build cache" (which is different then the normal caching it does for inputs and outputs for up-to-date checking.  It will not consider a task for caching in the build cache at all unless it is marked as `@Cacheable`.  And even then it will sometimes not be able to, although that is usually due to plugin dev errors or config errors in the consumer; but either way, it warns about these when they happen.

But you are absolutely correct that proper input and output definitions are critical for good build caching behavior.

**Comment by jskillin-idt** [comment]
> Not really. At least it won't make use of the "build cache" (which is different then the normal caching it does for inputs and outputs for up-to-date checking). It will not consider a task for caching in the build cache at all unless it is marked as `@Cacheable`. And even then it will sometimes not be able to, although that is usually due to plugin dev errors or config errors in the consumer; but either way, it warns about these when they happen.

Ahhhh! Fascinating! I've just been lucky to be using plugins that were written with this in mind. That's really good to know.

**Comment by jskillin-idt** [comment]
> As for build-caching and tests, I have no idea about what does that check. But considering `FROM-CACHE` is essentially the same as `SUCCESS`, can't that code just be changed?

I wrote an issue for this: https://github.com/quarkusio/quarkus/issues/25693

**Comment by sebersole** [comment]
Ok, deleting `~/.m2` fixed it.  

That can't be the full-time solution though, right?  @aloubyansky @glefloch - you guys don't do this each time do you?

**Comment by aloubyansky** [comment]
I never had to do it to run Gradle ITs, afair.

**Comment by sebersole** [comment]
> I never had to do it to run Gradle ITs, afair.

How do you verify changes you made to either of the plugins?

**Comment by aloubyansky** [comment]
Just `mvn -Dquickly` the plugin and then run IT tests or a sample project/reproducer.
Every time I rebase my branch I also do `mvn -Dquickly` before any dev/testing.

**Comment by aloubyansky** [comment]
That way I know my workspace as well as the local Maven repo is proper state.

**Comment by sebersole** [comment]
I'm not understanding "I never had to do it to run Gradle ITs" versus "and then run IT tests".  What is the difference?

**Comment by sebersole** [comment]
If I understand correctly, these `integration-tests/gradle/` tests are not in `devtools/gradle/...` because they rely on the other artifacts in the build being published to maven-local.  Is that accurate?

I guess that is pretty tough considering the overall build is handled by Maven.  We'd not really be able to set up a task dependency.  But man, does this cause a fugly situation.  It is simply a PITA testing any work one does on `devtools/gradle/`

**Comment by aloubyansky** [comment]
I never had to delete my local Maven repo for this. I built the whole repo after I rebase. That overrides previously installed snapshots in the local repo. Then I build and test modules I am working on.

**Comment by sebersole** [comment]
Could you give me some specifics?

> I built the whole repo after I rebase. 

What command do you use for this?

> Then I build and test modules I am working on.

What command here?  

I  want to update the README files to detail all these steps because they are quite involved and not documented; but honestly, I still have no idea what I am supposed to do (without deleting `~/.m2` e.g.).

Maybe y'all can help me understand this.  For the time being, I will only be working on these pieces of Gradleness in Quarkus.  Which seems to be particularly difficult given the overall build's use of Maven.   will be working on code under `devtools/gradle/` and sometimes under `integration-tests/gradle`.  So when I change code under `devtools/gradle/`, what are the complete steps I need to do in order to verify those changes (tests, codestyle, etc)?

I promise all of this will make it into the README for posterity ;)

**Comment by aloubyansky** [comment]
After a rebase from the root project dir `./mvnw -Dquickly`

After applying changes to a plugin `./mvnw -Dquickly` from the corresponding plugin module dir.

To run integration tests:
From the root project dir `cd integration-tests/gradle`
`./mvnw clean test -Dtest=<test-class-name>

**Comment by glefloch** [comment]
@sebersole I checked out your branch and built it locally. Tests are failing because a resolution of the `sourceDirectory` in the plugin extension [here](https://github.com/sebersole/quarkus/blob/9d19b638eca208b2c814dcd97dcd0a4f95f91b7c/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/extension/QuarkusPluginExtension.java#L65)

The error is: 
````
org.gradle.api.internal.provider.MissingValueException: Cannot query the value of task ':application:quarkusDev' property 'sourceDirectory' because it has no value available.
The value of this property is derived from:
  - extension 'quarkus' property 'sourceDirectory'
````

I tried to debug it, the [`lastDirectory`](https://github.com/sebersole/quarkus/blob/9d19b638eca208b2c814dcd97dcd0a4f95f91b7c/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/extension/QuarkusPluginExtension.java#L269) method always returns `null` as `fileSytemLocation` are not instance of `Directory` but instance of `DefaultFileSystemLocation` which is not a subtype of `Directory`. 

I don't know how we should fix that as one of your next commit will actually deprecate that.;.

**Comment by sebersole** [comment]
Ah, thanks for taking a look @glefloch !

I had some pressing tasks for Hibernate I needed to finish up last week, but today / tomorrow I will get back to this.

Now that I know about these separate integration tests I can hopefully check this stuff more incrementally.

As for how to continue... if y'all are ok with the sum total of the changes I made across all 3 of my pull requests (in theory/principle anyway) would you and @aloubyansky be agreeable to me simply consolidating them together and fix the combination against those integration tests?

**Comment by sebersole** [comment]
If not, I'll try to fix each individually but that's obviously more work.  And they build on each other anyway.

**Comment by glefloch** [comment]
I think it's ok as they will be part of the same version

**Comment by sebersole** [comment]
Superseded by https://github.com/quarkusio/quarkus/pull/25819

**Pull request closed without merging**